File: backend/src/main/java/org/akashihi/mdg/api/util/FilterConverter.java
package org.akashihi.mdg.api.util;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

public final class FilterConverter {
    private FilterConverter() {}

    public static Optional<Map<String, String>> buildFilter(Optional<String> query, ObjectMapper objectMapper) {
        return query.map(s -> {
            try {
                var queryMap = new HashMap<String,String>();
                var parsedQuery = objectMapper.readValue(s, Map.class);
                parsedQuery
                        .keySet().stream().filter(k -> k instanceof String && parsedQuery.get(k) instanceof String)
                        .forEach(k -> queryMap.put((String) k, (String) parsedQuery.get(k)));
                return queryMap;
            } catch (JsonProcessingException e) {
                return Collections.EMPTY_MAP;
            }
        });
    }
}

File: backend/src/main/kotlin/org/akashihi/mdg/api/util/FilterConverter.kt
package org.akashihi.mdg.api.util

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import java.util.Collections
import kotlin.collections.HashMap

object FilterConverter {
    fun buildFilter(query: String?, objectMapper: ObjectMapper): Map<String, String> {
        if (query == null) {
            return Collections.emptyMap()
        }
        val parsedQuery = try {
            objectMapper.readValue(query, Map::class.java)
        } catch (_: JsonProcessingException) {
            null
        } ?: return Collections.emptyMap()
        val queryMap = HashMap<String, String>()
        parsedQuery
            .keys.filter { it is String && parsedQuery[it] is String }
            .forEach { queryMap[it as String] = parsedQuery[it] as String }
        return queryMap
    }
}
